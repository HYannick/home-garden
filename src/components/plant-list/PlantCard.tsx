import React from 'react';
import { NavLink } from 'react-router-dom';
import styled from '@emotion/styled';
import Drop from '../../core/svg/Drop';

const Card: any = styled(NavLink)`
  display: block;
  position: relative;
  margin-bottom: 4rem;
`;

Card.Picture = styled('div')`
  position: relative;
  border-radius: 2rem;
  overflow: hidden;
  height: 12rem;
  > img {
    width: 100%;
    height: 100%;
    object-fit: cover;
  }
`;

Card.Overlay = styled('div')`
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-color: ${({ theme }) => theme.palette.grey.darkest};
  opacity: 0.65;
`;

Card.Infos = styled('div')`
  position: absolute;
  bottom: -2rem;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  > p {
    color: ${({ theme }) => theme.palette.light};
    margin: 0 0 1rem;
    font-size: 2rem;
    font-weight: 600;
  }
`;

// eslint-disable-next-line no-undef
Card.Chip = styled('div')<{ variant: string }>`
  display: flex;
  border: 0.4rem solid ${({ theme }) => theme.palette.light};
  background-color: ${({ theme, variant }) => theme.palette[variant || 'primary'].light};
  padding: 1rem 2rem;
  border-radius: 5rem;
  > span {
    color: ${({ theme, variant }) => theme.palette[variant || 'primary'].dark};
    margin-right: 1rem;
  }
  > svg {
    width: 2rem;
    height: 2rem;
    path {
      fill: none;
      stroke: ${({ theme, variant }) => theme.palette[variant || 'primary'].dark}
    }
  }
`;

export const setVariant = (days_left: number) => {
  let variant = 'primary';
  if(days_left < 3) {
    variant = 'warning';
  }

  if(days_left === 0) {
    variant = 'danger';
  }
  return variant;
};

const PlantCard: React.FC<any> = ({ plant }) => {

  return (
    <Card to={`/plants/${plant.id}`}>
      <Card.Picture>
        <img src={plant.picture} alt={plant.name}/>
        <Card.Overlay/>
      </Card.Picture>
      <Card.Infos>
        <p>{plant.name}</p>
        <Card.Chip variant={setVariant(plant.days_left)}>
          <span>Next Watering in {plant.days_left} days</span>
          <Drop/>
        </Card.Chip>
      </Card.Infos>
    </Card>
  );
};

export default PlantCard;
