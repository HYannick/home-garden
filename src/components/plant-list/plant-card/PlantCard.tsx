import React from 'react';
import Drop from '../../../core/svg/Drop';
import ImageFade from '../../image-fade/ImageFade';
import { setVariant } from '../../../core/utils/set_variant';
import CrossIcon from '../../../core/svg/Cross';
import { PlantCardTypes } from './PlantCard.types';
import { Card, CardWrapper } from './PlantCard.styled';

const PlantCard: React.FC<PlantCardTypes> = ({ t, plant, asSearchCard = false, path, onDelete }) => (
  <CardWrapper>
    {
      onDelete && (<Card.DeleteButton onClick={() => onDelete(plant.id)}>
        <CrossIcon/>
      </Card.DeleteButton>)
    }
    <Card to={path}>
      <Card.Picture asSearchCard={asSearchCard}>
        <ImageFade src={plant.picture} alt={plant.name} placeholder="#EFFFE2"/>js
        <Card.Overlay/>
      </Card.Picture>
      {
        !asSearchCard ? (
          <Card.Infos data-testid="default-card">
            <p>{plant.custom_name || plant.name}</p>
            <Card.Chip variant={setVariant(plant.days_left)}>
              <span>{t('plant_card.days_left', { days_left: plant.days_left })}</span>
              <Drop/>
            </Card.Chip>
          </Card.Infos>
        ) : (
          <Card.Infos data-testid="search-card">
            <Card.Chip variant="primary">
              <span>{plant.name}</span>
            </Card.Chip>
          </Card.Infos>
        )
      }
    </Card>
  </CardWrapper>
);

export default PlantCard;
