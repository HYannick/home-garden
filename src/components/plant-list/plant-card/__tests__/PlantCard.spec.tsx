import React from 'react';
import 'jest-dom/extend-expect';
import { cleanup } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { renderWithTheme } from '../../../../theme-wrapper';
import PlantCard from '../PlantCard';
import { PlantCardTypes } from '../PlantCard.types';

afterEach(cleanup);
describe('PlantCard', () => {
  const defaultProps: PlantCardTypes = {
    t: jest.fn().mockImplementation(() => 'plant_card.days_left'),
    plant: {
      name: 'Dracaena',
      custom_name: 'Draky',
      picture: null,
      days_left: 10,
    },
    asSearchCard: false,
    path: '',
  };

  it('should render properly', () => {
    const { container } = renderWithTheme(
      <Router><PlantCard {...defaultProps}/></Router>,
    );
    expect(container).toMatchSnapshot();
  });

  it('should render the default card template with day left', () => {
    const props = {
      ...defaultProps,
      asSearchCard: false,
    };
    const { getByTestId } = renderWithTheme(
      <Router><PlantCard {...props}/></Router>,

    );
    expect(getByTestId('default-card')).toBeTruthy();
  });

  it('should render the default search card template', () => {
    const props = {
      ...defaultProps,
      asSearchCard: true,
    };
    const { getByTestId } = renderWithTheme(
      <Router><PlantCard {...props}/></Router>,

    );
    expect(getByTestId('search-card')).toBeTruthy();
  });
});
