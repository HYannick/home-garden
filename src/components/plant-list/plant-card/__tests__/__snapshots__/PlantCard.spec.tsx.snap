// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlantCard should render properly 1`] = `
<div>
  <div
    class="css-13cuk55"
  >
    <a
      class="css-xtgc45"
      href="/"
    >
      <div
        class="css-gzgj0r"
      >
        <div
          class="css-prjb17"
          placeholder="#EFFFE2"
        >
          <img
            alt="Dracaena"
            class="css-10dr8mc"
          />
        </div>
        js
        <div
          class="css-go7c08"
        />
      </div>
      <div
        class="css-1egqyip"
        data-testid="default-card"
      >
        <p>
          Draky
        </p>
        <div
          class="css-1xu7jxi"
        >
          <span>
            plant_card.days_left
          </span>
          <svg
            viewBox="0 0 14.214 17"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6.944 1s6.27 6.2 6.27 9.329A5.9 5.9 0 0 1 7.107 16 5.9 5.9 0 0 1 1 10.329C1 7.193 6.944 1 6.944 1z"
              fill="#707070"
              stroke="#707070"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </div>
    </a>
  </div>
</div>
`;
